### PR TITLE
[bugfix] Fix MLA-793 Make Unity lifecycle methods protected.  Added tests for changes

### PR DESCRIPTION
**Pull request by surfnerd**
### Proposed change(s)

Address issue raised in #3579.

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)
#3579
https://jira.unity3d.com/browse/MLA-739

### Types of change(s)

- [x] Bug fix

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: Project/Project.sln.DotSettings [hunk 1]
```diff
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (chriselion) [inline Project/Project.sln.DotSettings]: Should this be here?

Comment (surfnerd) [reply Project/Project.sln.DotSettings]: It was added automatically by Rider. 🤷‍♂

File: com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs [hunk 2]
```diff
@@ -602,4 +603,56 @@ public void TestHeuristicPolicyStepsSensors()
...
+        public void TestAgentDontCallBaseOnEnable()
+        {
+            var go = new GameObject();
+            var agent = go.AddComponent<OnEnableAgent>();
+            var onEnable = typeof(OnEnableAgent).GetMethod("OnEnable", BindingFlags.NonPublic | BindingFlags.Instance);
+            var sendInfo = typeof(Agent).GetMethod("SendInfoToBrain", BindingFlags.NonPublic | BindingFlags.Instance);
+            Assert.NotNull(onEnable);
+            onEnable.Invoke(agent, null);
+            Assert.NotNull(sendInfo);
+            Assert.Throws<UnityAgentsException>(() =>
```
Comment (chriselion) [inline com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: I'm confused, why does this result in a `UnityAgentsException`?

Is it worth combining these to make it clearer that the setup is identical (and make it less likely they diverge in the future)?

Something like
```c#
void TestLifeCycleMethods(bool callbase)
{
  // do the setup
  if(!callbase) {
    Assert.Throws<..>...
  } else {
    Assert.DoesNotThrow...
  }
}

[Test]
public void TestAgentDontCallBaseOnEnable()
{
  TestLifeCycleMethods(false);
}

[Test]
public void TestAgentCallBaseOnEnable()
{
  TestLifeCycleMethods(true);
}

```

Comment (surfnerd) [reply com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: The Reflection Target Invoker catches the real exception, then throws its own.  So I catch the TargetInvocationException and throw the base exception (which is the one I'm expecting). Sure I can combine them.

Comment (chriselion) [reply com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: But what's the real one? I didn't see any `UnityAgentsException` thrown from `SendInfoToBrain()`

Comment (chriselion) [reply com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: Doh, I was looking locally, missed the new exception.

File: com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs [hunk 3]
```diff
@@ -602,4 +603,61 @@ public void TestHeuristicPolicyStepsSensors()
...
+    [TestFixture]
+    public class TestOnEnableOverride
+    {
+        public class OnEnableAgent : Agent
+        {
+            public bool callBase;
+
+            protected override void OnEnable()
+            {
+                if (callBase)
```
Comment (mmattar) [inline com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: missing { }

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 4]
```diff
@@ -570,6 +576,13 @@ internal void InitializeSensors()
         /// </summary>
         void SendInfoToBrain()
         {
+            if (!m_Initialized)
+            {
+                throw new UnityAgentsException("Call to SendInfoToBrain when Agent hasn't been initialized." +
+                    "Please ensure that you are calling 'base.OnEnable()' if you have overridden OnEnable.");
+
```
Comment (mmattar) [inline com.unity.ml-agents/Runtime/Agent.cs]: remove empty line

File: com.unity.ml-agents/CHANGELOG.md [hunk 5]
```diff
@@ -42,6 +42,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
  - `AgentInfo.actionMasks` has been renamed to `AgentInfo.discreteActionMasks`.
  - `DecisionRequester` has been made internal (you can still use the DecisionRequesterComponent from the inspector). `RepeatAction` was renamed `TakeActionsBetweenDecisions` for clarity. (#3555)
  - The `IFloatProperties` interface has been removed.
+ - Fix #3579.
```
Comment (mmattar) [inline com.unity.ml-agents/CHANGELOG.md]: Lets expand with a 1-liner and then have a link to the Issue

**Pull request merged**